File: src/design_patterns/cpp/iterator/Iterator.cpp
//
// Created by HuRF on 2022/9/21.
//

#include "Iterator.h"


Iterator *ConcreateAggregate::CreateIterator() {
    return new ConcreateIterator(this);
}

int ConcreateAggregate::GetCount() {
    return (int) items.size();
}

string ConcreateAggregate::operator[](int index) {
    if (this->items.size() <= index) {
        this->items.resize(index + 1);
    }
    return this->items[index];
}

void ConcreateAggregate::Insert(const string &strVal) {
    items.push_back(strVal);
}

ConcreateIterator::ConcreateIterator(ConcreateAggregate *aggregate) : aggregate(aggregate), current(0) {}

string ConcreateIterator::First() {
    return (*aggregate)[0];
}

string ConcreateIterator::Next() {
    string ret;
    current++;
    if (current < aggregate->GetCount()) {
        ret = (*aggregate)[current];
    }
    return ret;
}

bool ConcreateIterator::IsDone() {
    return current >= aggregate->GetCount();
}

string ConcreateIterator::CurrentItem() {
    return (*aggregate)[current];
}

ConcreateIteratorDesc::ConcreateIteratorDesc(ConcreateAggregate *aggregate) : aggregate(aggregate),
                                                                              current(aggregate->GetCount() - 1) {
}

string ConcreateIteratorDesc::First() {
    return (*aggregate)[aggregate->GetCount() - 1];
}

string ConcreateIteratorDesc::Next() {
    string ret;
    current--;
    if (current >= 0) {
        ret = (*aggregate)[current];
    }
    return ret;
}

bool ConcreateIteratorDesc::IsDone() {
    return current < 0;
}

string ConcreateIteratorDesc::CurrentItem() {
    return (*aggregate)[current];
}

File: src/design_patterns/python/iterator/iterator.py
# -*- coding:utf-8 -*-
'''
@File    :   iterator.py
@Time    :   2022/09/28 20:04:39
@Author  :   SheltonXiao
@Version :   1.0
@Contact :   pi620903@163.com
@Desc    :   None
'''


class Iterator(object):
    def first(self):
        pass
    def next(self):
        pass
    def is_done(self):
        pass
    def current_item(self):
        pass

class Aggregate(object):
    def create_iterator(self):
        pass

class ConcreteAggregate(Aggregate):
    def __init__(self):
        self.__items = []
    def __call__(self,index,value=None):
        if value != None:
            self.__items.insert(index,value)
        else:
            return self.__items[index]
    def create_iterator(self):
        return ConcreteIterator(self)
    def get_count(self):
        return len(self.__items) 
    count = property(get_count)

class ConcreteIterator(Iterator):
    __aggregate = None
    __current = 0
    def __init__(self,aggregate:ConcreteAggregate):
        self.__aggregate = aggregate
    def first(self):
        return self.__aggregate(0)
    def next(self):
        ret = None
        self.__current = self.__current + 1
        if self.__current < self.__aggregate.count:
            ret = self.__aggregate(self.__current)
        return ret
    def is_done(self):
        return self.__current >= self.__aggregate.count
    def current_item(self):
        return self.__aggregate(self.__current)

# 从后往前的迭代器
class ConcreteIteratorDesc(Iterator):
    __aggregate = None
    __current = 0
    def __init__(self,aggregate:ConcreteAggregate):
        self.__aggregate = aggregate
        self.__current = aggregate.count - 1
    def first(self):
        return self.__aggregate(self.__aggregate.count - 1)
    def next(self):
        ret = None
        self.__current = self.__current - 1
        if self.__current > 0:
            ret = self.__aggregate(self.__current)
        return ret
    def is_done(self):
        return self.__current < 0
    def current_item(self):
        return self.__aggregate(self.__current)


if __name__ == '__main__':
    a = ConcreteAggregate()
    a(0,"大鸟")
    a(1,"小菜")
    a(2,"行李")
    a(3,"老外")
    a(4,"公交内部员工")
    a(5,"小偷")

    #i = ConcreteIterator(a)
    i = ConcreteIteratorDesc(a)

    item = i.first()
    while not i.is_done():
        print(f"{i.current_item()} 请买车票！")
        i.next()
